# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
#функция преобразования среднего значения в %
def percent_mean(m):
    return '{:.3%}'.format(m.mean())

In [23]:
data.groupby('children')['debt'].agg(['count', percent_mean]).sort_values('percent_mean') 
# Ваш код будет здесь. Вы можете создавать новые ячейки.

count percent_mean
children                    
5             9       0.000%
0         14091       7.544%
3           330       8.182%
1          4808       9.235%
2          2052       9.454%
4            41       9.756%

**Вывод:**

Можно сделать вывод что есть зависимость между количеством детей и возвратом кредитов в срок. В семьях без детей самый низкий процент невозвратов по кредиту в срок, в то время как в семьях где есть дети процент выше (чем больше детей, тем выше процент просрочек). В семьях с 5 детьми сложно сделать однозначный вывод потому что таких семей взявших кредит очень мало.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
data.groupby(['family_status'])['debt'].agg(['count', percent_mean]).sort_values('percent_mean')
# Ваш код будет здесь. Вы можете создавать новые ячейки.

count percent_mean
family_status                            
вдовец / вдова           951       6.625%
в разводе               1189       7.065%
женат / замужем        12261       7.561%
гражданский брак        4134       9.313%
Не женат / не замужем   2796       9.764%

**Вывод:** 

Зависимость между семейным положением и возвратом кредита в срок есть. У клиентов, которые не были в браке процент просрочки по кредиту выше, чем у тех кто находится или был в браке.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
data.groupby('total_income_category')['debt'].agg(['count', percent_mean]).sort_values('percent_mean') 
# Ваш код будет здесь. Вы можете создавать новые ячейки.

count percent_mean
total_income_category                    
D                        349       6.017%
B                       5014       7.060%
A                         25       8.000%
C                      15921       8.498%
E                         22       9.091%

**Вывод:** 

Сложно сказать наверняка о том зависит ли уровень дохода с возвратом кредитов в срок, потому что недостаточно информации из-за низкого размера групп (D, A и Е).

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
data.groupby('purpose_category')['debt'].agg(['count', percent_mean]).sort_values('percent_mean') 
# Ваш код будет здесь. Вы можете создавать новые ячейки.

count percent_mean
purpose_category                            
операции с недвижимостью  10751       7.255%
проведение свадьбы         2313       7.912%
получение образования      3988       9.253%
операции с автомобилем     4279       9.348%

**Вывод:** 

Кредиты оформленные на операции с недвижимостью имеют самый низкий процент просрочки (7.255%), в то время как кредиты выданные на операции с автомобилем имеют наоборот самый выскоий процент просрочки (9.348%). Кредиты выданные на получение образования также имеют достаточно выский процент просрочки (9.253%).

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Возможно пропуски появились случайно. Банк не стал бы выдавать кредит при отсутствии каких-либо данных из таблицы.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медиальные значения более усточивы к выбросам (слишком большие или наоборот слишком маленькие значения) и аномальным значениям. 

### Шаг 4: общий вывод.

Во время выполнения исследования была изучена общая информация по заемщикам, удалены пропуски в таблице, обработаны аномальные значения и дубликаты, также клиенты были объединены в группы по разным параметрам. По итогу проделанной работы были даны ответы на следующие вопросы:
1.Есть ли зависимость между количеством детей и возвратом кредита в срок?
2.Есть ли зависимость между семейным положением и возвратом кредита в срок?
3.Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4.Как разные цели кредита влияют на его возврат в срок?

Также были даны обьяснения на вопросы не относящиеся к самому исследованию.

В ходе исследования была изучена общая информация по заемщикам, удалены пропуски в таблице, обработаны аномальные значения и дубликаты, также клиенты были объединены в группы по разным параметрам. По итогу проделанной работы были даны ответы на следующие вопросы:

1. **Есть ли зависимость между количеством детей и возвратом кредита в срок?** 
    
    * В семьях где нет детей самый низкий процент просрочек по кредитам - 7,54%.
    * В семьях с 1 или 2 детьми 9.23% и 9.45% соответственно.
    * В семьях с 4 детьми самый высокий процент по просроченным кредитам - 9.75%.  
    
    Как видно зависимость между количеством детей и возвратом кредитов есть прямая связь, потому что больше семей без детей выплачивают кредиты в срок в отличии от семей где есть дети.Скорее всего это связано с тем, что в семьях без детей меньше непредвиденных расходов и такие семьи могут планировать свой бюджет более чётко, чем семьи с детьми. Семьи с 5 детьми не учитывались в подсчетах из-за количества таких семей сложно сделать однозначный вывод. 
    
    
2. **Есть ли зависимость между семейным положением и возвратом кредита в срок?**
    
    Зависимость между семейным положением и возвратом кредита в срок есть. У клиентов, которые не были в браке процент
    просрочки по кредиту выше, чем у тех кто находится или был в браке:
    * Овдовевшие имеют самый низкий процент по просрочке - 6.62%.
    * Люди находящиеся в разводе - 7.06%.
    * Женатые и замужние - 7.56%.
    * Состоящие в гражданском браке - 9.31%.
    * Не состоящие в браке имеют самыый высокий процент просрочки - 9.76%.  
    
    Скорее всего это связано с тем, что люди состоящие или состоявшие в браке более ответственно относятся к выплатам и имеют более высокий суммарный доход, чем люди которые никогда не состояли в браке. Вдовы и вдовцы скорее всего имеют самый низкий процент по просрочке в силу того что это как правило возростные люди, которые относятся к долгам с максимальной серьезностью и им в силу возраста скорее всего помогают с выплатами взрослые дети.
    
    
3. **Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**
    
    Сложно судить наверняка о том зависит ли уровень дохода с возвратом кредитов в срок, потому что недостаточно информации из-за низкого размера групп.
    
    
4. **Как разные цели кредита влияют на его возврат в срок?**
    
    * Кредиты оформленные на операции с недвижимостью имеют самый низкий процент просрочки - 7.25%.
    * У кредитов на свадьбу процент просрочки так же достаточно низкий - 7.91%.
    * Самый высокий процент просрочки  по крдитам выданым на операции с автомобилем - 9.34% .
    * Кредиты выданные на получение образования также имеют достаточно выский процент просрочки - 9.25%.  
    
    
    
По итогу проведенного исследования можно сделать следующие выводы:

1. Семьи без детей реже просрачивают кредит, чем семьи с детьми. При чем чем больше детей, тем чаще кредит не выплачивают в срок.
2. Люди бывшие в браке платят в срок чаще тех, кто в браке не был. 
3. От того на что взят кредит зависит и его возврат в срок.